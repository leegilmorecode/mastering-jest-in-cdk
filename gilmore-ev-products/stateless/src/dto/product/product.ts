type EVProductCategory = 'Vehicle' | 'Battery' | 'Charger' | 'Accessory';

interface EVProductBase {
  id?: string;
  name: string;
  description: string;
  price: number;
  category: EVProductCategory;
  tags?: string[];
  available: boolean;
  releaseDate?: Date;
  created?: string;
  updated?: string;
}

interface VehicleSpecifications {
  range: number; // in kilometers or miles
  batteryCapacity: number; // in kWh
  topSpeed: number; // in km/h or mph
  acceleration: number; // 0-100 km/h or 0-60 mph time
  seatingCapacity: number;
  drivetrain: 'FWD' | 'RWD' | 'AWD'; // Front-Wheel Drive, Rear-Wheel Drive, All-Wheel Drive
}

interface BatterySpecifications {
  capacity: number; // in kWh
  type: 'Lithium-Ion' | 'Solid-State' | 'Other';
  weight: number; // in kilograms
  voltage: number; // in volts
  dimensions: {
    length: number; // in centimeters
    width: number; // in centimeters
    height: number; // in centimeters
  };
}

interface ChargerSpecifications {
  powerOutput: number; // in kW
  type: 'Level 1' | 'Level 2' | 'DC Fast Charger';
  connectorType: 'Type 1' | 'Type 2' | 'CCS' | 'CHAdeMO' | 'Tesla';
}

interface AccessorySpecifications {
  compatibility: string[]; // List of compatible vehicles or products
  weight?: number; // in kilograms
  dimensions?: {
    length: number; // in centimeters
    width: number; // in centimeters
    height: number; // in centimeters
  };
}

export type Product =
  | (EVProductBase & {
      category: 'Vehicle';
      specifications: VehicleSpecifications;
    })
  | (EVProductBase & {
      category: 'Battery';
      specifications: BatterySpecifications;
    })
  | (EVProductBase & {
      category: 'Charger';
      specifications: ChargerSpecifications;
    })
  | (EVProductBase & {
      category: 'Accessory';
      specifications: AccessorySpecifications;
    });
