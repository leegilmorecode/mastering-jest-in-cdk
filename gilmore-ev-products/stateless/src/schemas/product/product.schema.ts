export const schema = {
  $schema: 'http://json-schema.org/draft-07/schema#',
  type: 'object',
  properties: {
    id: {
      type: 'string',
    },
    name: {
      type: 'string',
    },
    description: {
      type: 'string',
    },
    price: {
      type: 'number',
    },
    category: {
      type: 'string',
      enum: ['Vehicle', 'Battery', 'Charger', 'Accessory'],
    },
    tags: {
      type: 'array',
      items: {
        type: 'string',
      },
    },
    available: {
      type: 'boolean',
    },
    releaseDate: {
      type: 'string',
      format: 'date-time',
    },
    created: {
      type: 'string',
      format: 'date-time',
    },
    updated: {
      type: 'string',
      format: 'date-time',
    },
    specifications: {
      oneOf: [
        {
          type: 'object',
          properties: {
            range: {
              type: 'number',
            },
            batteryCapacity: {
              type: 'number',
            },
            topSpeed: {
              type: 'number',
            },
            acceleration: {
              type: 'number',
            },
            seatingCapacity: {
              type: 'integer',
            },
            drivetrain: {
              type: 'string',
              enum: ['FWD', 'RWD', 'AWD'],
            },
          },
          required: [
            'range',
            'batteryCapacity',
            'topSpeed',
            'acceleration',
            'seatingCapacity',
            'drivetrain',
          ],
          additionalProperties: false,
        },
        {
          type: 'object',
          properties: {
            capacity: {
              type: 'number',
            },
            type: {
              type: 'string',
              enum: ['Lithium-Ion', 'Solid-State', 'Other'],
            },
            weight: {
              type: 'number',
            },
            voltage: {
              type: 'number',
            },
            dimensions: {
              type: 'object',
              properties: {
                length: {
                  type: 'number',
                },
                width: {
                  type: 'number',
                },
                height: {
                  type: 'number',
                },
              },
              required: ['length', 'width', 'height'],
              additionalProperties: false,
            },
          },
          required: ['capacity', 'type', 'weight', 'voltage', 'dimensions'],
          additionalProperties: false,
        },
        {
          type: 'object',
          properties: {
            powerOutput: {
              type: 'number',
            },
            type: {
              type: 'string',
              enum: ['Level 1', 'Level 2', 'DC Fast Charger'],
            },
            connectorType: {
              type: 'string',
              enum: ['Type 1', 'Type 2', 'CCS', 'CHAdeMO', 'Tesla'],
            },
          },
          required: ['powerOutput', 'type', 'connectorType'],
          additionalProperties: false,
        },
        {
          type: 'object',
          properties: {
            compatibility: {
              type: 'array',
              items: {
                type: 'string',
              },
            },
            weight: {
              type: 'number',
            },
            dimensions: {
              type: 'object',
              properties: {
                length: {
                  type: 'number',
                },
                width: {
                  type: 'number',
                },
                height: {
                  type: 'number',
                },
              },
              additionalProperties: false,
            },
          },
          required: ['compatibility'],
          additionalProperties: false,
        },
      ],
    },
  },
  required: [
    'id',
    'name',
    'description',
    'price',
    'category',
    'available',
    'created',
    'updated',
    'specifications',
  ],
  additionalProperties: false,
};
