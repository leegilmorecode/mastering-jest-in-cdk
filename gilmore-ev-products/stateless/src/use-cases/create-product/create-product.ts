import { getISOString, logger, schemaValidator } from '@shared';

import { Product } from '@dto/product';
import { ValidationError } from '@errors/validation-error';
import { config } from '@config';
import { schema } from '@schemas/product';
import { upsert } from '@adapters/secondary/dynamodb-adapter';
import { v4 as uuid } from 'uuid';

const tableName = config.get('tableName');

export async function createProductUseCase(
  newProduct: Product
): Promise<Product> {
  try {
    const createdDate = getISOString();
    const id = uuid();

    const product: Product = {
      ...newProduct,
      id,
      created: createdDate,
      updated: createdDate,
    };

    schemaValidator(schema, product);

    // Note: we add this to show some business logic to unit test
    if (product.price > 10000 && product.category === 'Charger') {
      throw new ValidationError(
        `Product price of ${product.price} is out of range`
      );
    }

    await upsert<Product>(product, tableName, id);

    return product;
  } catch (error) {
    logger.error(`Create Product ${error}`);
    throw error;
  }
}
