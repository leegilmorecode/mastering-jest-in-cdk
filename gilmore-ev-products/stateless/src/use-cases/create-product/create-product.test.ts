import { Product } from '@dto/product';
import { createProductUseCase } from './create-product';
import { logger } from '@shared/logger';
import { upsert } from '@adapters/secondary/dynamodb-adapter';

jest.mock('@shared/logger');
jest.mock('@adapters/secondary/dynamodb-adapter');

let newProduct: Product;

describe('create-product-use-case', () => {
  beforeAll(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2024, 1, 1));
  });

  beforeEach(() => {
    jest.clearAllMocks();

    newProduct = {
      name: 'QuickCharge Station',
      description: 'Fast charging station for electric vehicles',
      price: 2000,
      category: 'Charger',
      available: true,
      specifications: {
        powerOutput: 150,
        type: 'DC Fast Charger',
        connectorType: 'CCS',
      },
    };
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('should create a product successfully', async () => {
    // arrange / act
    (upsert as jest.Mock).mockResolvedValue(newProduct);
    const product = await createProductUseCase(newProduct);

    // assert
    expect(product).toMatchSnapshot();
  });

  it('should log an error on failure', async () => {
    expect.assertions(2);

    // arrange
    const error = new Error('Unhandled error');
    (upsert as jest.Mock).mockRejectedValue(error);

    // act / assert
    await expect(createProductUseCase(newProduct)).rejects.toThrow(
      'Unhandled error'
    );
    expect(logger.error).toHaveBeenCalledWith(
      'Create Product Error: Unhandled error'
    );
  });

  it('should throw an error if price is over Â£10K in category charger', async () => {
    // arrange
    newProduct.price = 200000; // Note: in our business logic this can only be 10K max for category charger

    // act / assert
    await expect(createProductUseCase(newProduct)).rejects.toThrow(
      'Product price of 200000 is out of range'
    );
  });
});
