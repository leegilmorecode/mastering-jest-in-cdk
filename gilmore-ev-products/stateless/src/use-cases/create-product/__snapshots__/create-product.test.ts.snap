// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-product-use-case should create a product successfully 1`] = `
{
  "available": true,
  "category": "Charger",
  "created": "2024-02-01T00:00:00.000Z",
  "description": "Fast charging station for electric vehicles",
  "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "name": "QuickCharge Station",
  "price": 2000,
  "specifications": {
    "connectorType": "CCS",
    "powerOutput": 150,
    "type": "DC Fast Charger",
  },
  "updated": "2024-02-01T00:00:00.000Z",
}
`;
